components/tidb_query_vec_executors/src/index_scan_executor.rs

```diff
@@ -232,59 +321,147 @@ impl IndexScanExecutorImpl {
-                self.decode_handle_from_value(&value[value.len() - 8..])?
-            } else {
-                // This is a normal index. The remaining payload part is the PK handle.
-                // Let's decode it and put in the column.
-                self.decode_handle_from_key(&key[key.len() - 9..])?
-            };
-            columns[self.columns_id_without_handle.len()]
-                .mut_decoded()
-                .push_int(Some(handle_val));
+    fn process_unique_common_handle_value(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        let handle_len = (&value[2..]).read_u16().map_err(|_| {
+            other_err!(
+                "Fail to read common handle's length from value: {:?}",
+                value
+            )
+        })? as usize;
+        let handle_end_offset = 4 + handle_len;
+
+        if handle_end_offset > value.len() {
+            return Err(other_err!("`handle_len` is corrupted: {}", handle_len));
         }
+
+        // If there are some restore data.
+        if handle_end_offset < value.len() {
+            let restore_values = &value[handle_end_offset..];
+            self.extract_columns_from_row_format(restore_values, columns)?;
+        } else {
+            // The datum payload part of the key.
+            let mut key_payload = &key[table::PREFIX_LEN + table::ID_LEN..];
+            Self::extract_columns_from_datum_format(
+                &mut key_payload,
+                &mut columns[..self.columns_id_without_handle.len()],
+            )?;
+        }
+
+        let mut common_handle = &value[4..handle_end_offset];
+        Self::extract_columns_from_datum_format(
+            &mut common_handle,
+            &mut columns[self.columns_id_without_handle.len()..],
+        )?;
+
         Ok(())
     }
 
-    fn process_kv_pair_old(
+    fn process_normal_new_collation_value(
+        &mut self,
+        mut key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        use DecodeHandleStrategy::*;
+        let tail_len = value[0] as usize;
+
+        if tail_len > value.len() {
+            return Err(other_err!("`tail_len`: {} is corrupted", tail_len));
+        }
+
+        let restore_values = &value[1..value.len() - tail_len];
+        self.extract_columns_from_row_format(restore_values, columns)?;
+
+        match self.decode_handle_strategy {
+            NoDecode => {}
+            DecodeIntHandle if tail_len < 8 => {
+                key = &key[table::PREFIX_LEN + table::ID_LEN..];
+                datum::skip_n(&mut key, self.columns_id_without_handle.len())?;
+                let handle = self.decode_handle_from_key(key)?;
+                columns[self.columns_id_without_handle.len()]
+                    .mut_decoded()
+                    .push_int(Some(handle));
+            }
+            DecodeIntHandle => {
```

Comment:
In what case, will a normal index has int handle in the value?

Reply:
More comments have been added.